test/unit/spec/util/backoff.js

```diff
@@ -0,0 +1,123 @@
+'use strict';
+
+const DefaultBackoff = require('../../../../lib/util/backoff');
+const sinon = require('sinon');
+
+describe('Backoff', () => {
+  describe('Function Call', () => {
+    let options = {};
+    let fn;
+    let backoff;
+    let fakeTimer;
+    beforeEach(() => {
+      fn = sinon.spy();
+      backoff = new DefaultBackoff(options);
+      fakeTimer = sinon.useFakeTimers();
+    });
+
+    afterEach(() => {
+      fn.resetHistory();
+      backoff.reset();
+      fakeTimer.restore();
+    });
+
+    it('should call the function on start', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(100);
+      backoff.reset();
+      sinon.assert.calledOnce(fn);
+    });
+
+    it('should increase the duration exponentially', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(110);
+      backoff.backoff(fn);
+      fakeTimer.tick(210);
+      backoff.backoff(fn);
+      fakeTimer.tick(410);
+      sinon.assert.calledThrice(fn);
+    });
+
+    it('should reset the duration', () => {
+      backoff.backoff(fn);
+      fakeTimer.tick(110);
+      backoff.reset();
+      fakeTimer.tick(210);
+      fakeTimer.tick(410);
+      sinon.assert.calledOnce(fn);
+    });
+  });
+
+  describe('Options', () => {
+    let fn;
+    let fakeTimer;
+    beforeEach(() => {
+      fn = sinon.spy();
+      fakeTimer = sinon.useFakeTimers();
+    });
+
+    afterEach(() => {
+      fn.resetHistory();
+      fakeTimer.restore();
+    });
+
+    it('min', () => {
+      const options = {
+        min: 10,
+      };
+      const backoff = new DefaultBackoff(options);
+
+      backoff.backoff(fn);
+      fakeTimer.tick(10);
+      backoff.reset();
+      sinon.assert.calledOnce(fn);
+    });
+
+    it('max', () => {
+      const options = {
+        max: 1000,
+      };
+      const backoff = new DefaultBackoff(options);
+
+      backoff.backoff(fn);
+      fakeTimer.tick(100);
+      backoff.backoff(fn);
+      fakeTimer.tick(200);
+      backoff.backoff(fn);
+      fakeTimer.tick(400);
+      backoff.backoff(fn);
+      fakeTimer.tick(800);
+      sinon.assert.callCount(fn, 4);
+      backoff.reset();
```

Comment:
looks like this test isn't really verifying 'max' property since you call reset here.

Reply:
Hm, since we changed how our backoff behaves, I'm thinking it's going to be hard since we don't call a function automatically (onTimer) until the max duration. We're now using max as a variable for our setTimeout duration calculation so that we don't get numbers beyond the max that was set. 

So you're correct, but I'm thinking if there's a better way to test "max" since it's no longer being used to determine if we should stop retrying, and used more simply as a means to calculate a new duration once `backoff` is called. But `_duration` is a private member of this class so it doesn't seem right to check the duration? WDYT?

Reply:
I think way you could verify max is to write a test that would tick at 10, 20, 40, 80, 160, 
and verify that at tickCount = 140 you get 4 calls without max, and 5 calls with max set to 40.

Reply:
We currently have backoff implemented so that we only start the operation when we call `backoff`. It's no longer calling on another function to continuously retry the function. 

So for example this test:
`    it('max', () => {
      const options = {
        max: 40,
      };
      const backoff = new DefaultBackoff(options);

      backoff.backoff(fn);
      fakeTimer.tick(10);
      fakeTimer.tick(20);
      fakeTimer.tick(40);
      fakeTimer.tick(160);
      sinon.assert.callCount(fn, 1);
      backoff.reset();
    });`
    
    This will pass because we only called `backoff` once. Basically, the max duration is no longer being used to determine our attempts at calling a function anymore. It's only used to determine the max duration for our setTimeout